File: 2024/13_claw_contraption.jl
example = split("Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=8400, Y=5400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=12748, Y=12176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=7870, Y=6450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=18641, Y=10279", '\n')


mutable struct Game
    dxa::Int
    dya::Int
    dxb::Int
    dyb::Int
    zx::Int
    zy::Int
end  


function solve_game(game::Game)::Int
    #=
    We will solve a game by solving the linear equation A * na + B * nb = z
    while minimising cost = 3 * na + nb. But, there can be only one solution,
    so we need not worry about minimising.

    Actually, we can directly solve this equation system:
      | dxa * na + dxb * nb = zx
      | dya * na + dyb * nb = zy

      => na = 1 / dxa * (zx - dxb * nb)
      => nb = (zy - zx * dya / dxa) / (dyb - dxb * dya / dxa)
    =#
    dxa, dya, dxb, dyb, zx, zy = game.dxa, game.dya, game.dxb, game.dyb, game.zx, game.zy

    nb = Int(round((zy - zx * dya / dxa) / (dyb - dxb * dya / dxa)))
    na = Int(round((zx - dxb * nb) / dxa))

    # Check that we are actually hitting the target:
    tx = dxa * na + dxb * nb
    ty = dya * na + dyb * nb
    
    if abs(tx - zx) == 0 && abs(ty - zy) == 0
        return Int(round(3 * na + nb))
    else
        return 0
    end
end


function parse_inp(inp, offset=0)
    games = []

    a, b, z = Vector{Int}, Vector{Int}, Vector{Int}
    for line in inp
        if contains(line, "Button A")
            a = [parse(Int, x.match) for x in eachmatch(r"\d+", line)]
        elseif contains(line, "Button B")
            b = [parse(Int, x.match) for x in eachmatch(r"\d+", line)]
        elseif contains(line, "Prize")
            z = [parse(Int, x.match) for x in eachmatch(r"\d+", line)]
        elseif length(line) == 0
            push!(games, Game(a[1], a[2], b[1], b[2], z[1] + offset, z[2] + offset))
        end
    end
    push!(games, Game(a[1], a[2], b[1], b[2], z[1] + offset, z[2] + offset))
    return games
end


@time begin
    for (k, offset) in enumerate([0, 10000000000000])
        for (name, games) in [("example", parse_inp(example, offset)), ("puzzle", parse_inp(readlines("data/13.dat"), offset))]
            println("Part $k, $name: ", sum(solve_game(game) for game in games))
        end
        println()
    end
end

File: 2024/13_claw_contraption.py
from aocd import get_data
import re
import numpy as np

example = """Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=8400, Y=5400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=12748, Y=12176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=7870, Y=6450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=18641, Y=10279""".splitlines()

dat = get_data(year=2024, day=13).splitlines()


def parse(data):
    blocks = []
    block = []
    for line in data:
        if len(line) == 0:
            blocks.append(block[:])
            block = []
        else:
            block.append(line)
    blocks.append(block)

    for k, block in enumerate(blocks):
        A = [int(x) for x in re.findall(r'\d+', block[0])]
        B = [int(x) for x in re.findall(r'\d+', block[1])]
        target = np.asarray([int(x) for x in re.findall(r'\d+', block[2])])
        M = np.asarray([[A[0], B[0]], [A[1], B[1]]])
        blocks[k] = (M, target)

    return blocks

def solve(data, offset=0):
    res = 0
    for k, (M, target) in enumerate(parse(data)):
        target += offset
        presses = np.dot(np.linalg.inv(M), target)
        if presses is not None:
            A, B = [round(x) for x in presses]
            test = np.dot(M, [A, B])

            if abs(test[0] - target[0]) < 1 and abs(test[1] - target[1]) < 1:
                res += 3 * A + B

    return res

if solve(example) == 480:
    print("Part 1: ", solve(dat))
    print("Part 2: ", solve(dat, offset=10000000000000))
